src/http-api.js

```diff
@@ -83,6 +83,44 @@ module.exports.MatrixHttpApi = function MatrixHttpApi(event_emitter, opts) {
     this.uploads = [];
 };
 
+/**
+ * Construct a Timeout Context.
+ * Set onProgressCheck to receive progressEvent from an XMLHTTPRequest.
+ * TimeoutContext will debounce the timeoutFunction for waitMs if progress
+ * is being made.
+ *
+ * @constructor
+ * @param {Function} timeoutFunction The function to debounce with waitMs.
+ * @param {Number} waitMs The time to debounce the callback in miliseconds.
+ *
+ */
+const TimeoutContext = function TimeoutContext(timeoutFunction, waitMs) {
+    this.lastLoaded = null;
+    this.jobName = null;
+    this.timeoutFunction = timeoutFunction;
+    this.waitMs = waitMs;
+};
+
+// For testing
+module.exports.__TimeoutContext = TimeoutContext;
+
+TimeoutContext.prototype.onProgressCheck = function(progressEvent) {
+    if (this.lastLoaded || progressEvent.loaded > this.lastLoaded) {
```

Comment:
Why are you comparing Numbers to `null` initially? Why isn't `lastLoaded` set to `0`? Also, you should comment why you are debouncing for any previous positive number. Isn't `progressEvent.loaded > this.lastLoaded` good enough? One can make the argument that if the request has wedged at say `20582` loaded bytes, and isn't changing, then you should NOT be debouncing and waiting forever if we are not making any progress.